nx/lib/nx.ex

```diff
@@ -5994,8 +6006,12 @@ defmodule Nx do
       Nx.Shared.raise_complex_not_supported("complex", 2)
     end
 
+    {_, realprecision} = type(real)
+    {_, imagprecision} = type(imag)
+    precision = Kernel.max(64, Kernel.min(realprecision, imagprecision) * 2)
```

Comment:
You should use Nx.Type.merge_type in combination with Nx.Type.to_real to get the correctly promoted precision.

This is probably why s64 is getting cast to c128 instead of c64

Reply:
Thanks! I had not seen the `Nx.Type` module. Using the functions from that module makes this much better and fixes the issue with `:s64` that you noted.